### ISSUE TITLE
Entity Id gets exposed in PanacheEntity in OpenAPI schema / Swagger UI for POST/PUT methods

### ISSUE THREAD

**Issue by rsvoboda** [quarkusio/quarkus#40006]
### Describe the bug

Entity Id gets exposed in PanacheEntity in OpenAPI schema / Swagger UI for POST/PUT methods

PanacheEntity defines `id` - https://github.com/quarkusio/quarkus/blob/main/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheEntity.java#L35

When people experimenting with the Swagger UI and sending posts as shown in the UI they get error because `id` is generated. The error isn't too intuitive if people do not have some context, it's just saying something like `Caused by: jakarta.persistence.EntityExistsException: detached entity passed to persist: my.foo.Entity`.

**Ask:** Panache* is quite pushed by Quarkus team, the OOTB experience with Swagger UI should be improve in my opinion.

If I don't use PanacheEntity I can get around this using `org.eclipse.microprofile.openapi.annotations.media.Schema` annotation
```java
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Schema(required = true, readOnly = true)
    public Long id;
```
When using this approach I can properly use SwaggerUI o experiment with the POST endpoints.

Screenshot from hibernate-orm-rest-data-panache-quickstart with added openapi extension:
<img width="1104" alt="Screenshot 2024-04-11 at 9 17 55" src="https://github.com/quarkusio/quarkus/assets/925259/8652c3f6-dab2-4fb6-bc7d-20e748b2a254">

### Expected behavior

Entity Id doesn't get exposed in PanacheEntity in OpenAPI schema / Swagger UI for POST/PUT methods

### Actual behavior

Entity Id gets exposed in PanacheEntity in OpenAPI schema / Swagger UI for POST/PUT methods

### How to Reproduce?

 - Use hibernate-orm-rest-data-panache-quickstart from https://github.com/quarkusio/quarkus-quickstarts
 - Add io.quarkus:quarkus-smallrye-openapi to the pom.xml
 - Run the application in dev mode
 - Open http://localhost:8080/q/swagger-ui/#/PeopleResource/post_my_people
 - Hit `Try it out` button on the right, and then click on `Execute` button
 - You will receive response code 500 and error in the application log

### Output of `uname -a` or `ver`

macOS

### Output of `java -version`

Java 17

### Quarkus version or git rev

Quarkus main

**Comment by mkouba** [comment]
> If I don't use PanacheEntity I can get around this using `org.eclipse.microprofile.openapi.annotations.media.Schema` annotation

Yes, you can still extend the `PanacheEntityBase` and define a custom id field (as described [in the docs](https://quarkus.io/guides/hibernate-orm-panache#custom-ids)). But it's not very convenient. Hopefully, there's a way to hook into the OpenAPI processing because I don't think that we can use Quarkus-specific stuff here (such as `AnnotationsTransformer`) :shrug:.

**Comment by phillip-kruger** [comment]
I think you can use a Filter. See https://github.com/quarkusio/quarkus/blob/5871b82bae238402055885689611d943344d4fde/extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java#L306C41-L306C72 as an example usage. (This one Health endpoints are added, but your filter can do anything)

**Comment by mkouba** [comment]
> Filter ... in hibernate-orm-panache extension?

I think rather in the `hibernate-orm-rest-data-panache` extension...

**Comment by phillip-kruger** [comment]
The filter build item is in a spi in openAPI specific for this usage yes. You would use it somewhere in panache hibernate.

**Comment by FroMage** [comment]
We already add Jackson and JSONB annotations in Panache to make some fields like `EntityManager` ignored in JSON/XML. We could add `Schema` as well, frankly.

**Comment by phillip-kruger** [comment]
Interesting. OpenAPI supports Jackson and JSONB. This might be a case of the example generated not using those. @MikeEdgar  ?

**Comment by MikeEdgar** [comment]
We should use them (Jackson/JsonB annotations) if they're present. They may not be set in a way that results in a read-only property though.

**Comment by FroMage** [comment]
Well, they're set on `getEntityManager` not on the `id` field. Because they target serialisation, not Hibernate behaviour.

So, in most cases, we want the `id` to be serialised to JSON, because it is relevant. Just like for `PUT` it is also relevant to deserialise it, especially in cases where it's allowed to update the `id`. For `POST` probably not.

**Comment by MikeEdgar** [comment]
I see. So marking the `id` as `readOnly` via `@Schema` may be the only option here. Is there any case where Panache allows an `id` to be altered by the `PUT` operation?

**Comment by FroMage** [comment]
Not via the `hibernate-orm-rest-data-panache` extension, no. But these entities are also used by other users, and it's possible one will expect the `id` field to be populated on deserialisation. If only to call `entityManager.merge(entity)` and get an attached version, or `EntityType.findById(entity.id)`, whatever variation.

As for being able to alter an ID, JPA says it's not allowed, or at least not portable:

> The application must not change the value of the primary key of an entity instance after the instance is made persistent

From https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2-m2#a132

Making the entity ID `readOnly` might affect other applications using entities as request bodies in `quarkus-rest` and expecting that field to be set via JSON deserialisation.

I'm not sure how many there are, if they exist.

**Comment by nicolasduminil** [comment]
Using `org.eclipse.microprofile.openapi.annotations.media.Schema` annotation, as follows:

      @Id
      @GeneratedValue
      @Schema(required = true, readOnly = true)
      private Long id;

still raises the mentioned exception, as shown in the attached screen copy. Any solution to this issue since ?
![Screenshot_20240911_191245](https://github.com/user-attachments/assets/603567db-c2bd-42f9-87aa-06cff70f4d90)

**Comment by FroMage** [comment]
Well `@Schema` is only useful for OpenAPI to generate examples. If you do send an ID for `POST` you will get the same error, for sure. The point was that if we use that, OpenAPI would not include an ID in the `POST` examples.

Now that I think about it, this appears to be an issue in rest-data-panache, which uses the entities as DTOs, which is something we never advise people to do,and in fact this only works in the most trivial cases.

We don't want to add `@Schema` to the `PanacheEntity.id` field because there are lots of legitimate use-cases for sending IDs along with entities to `POST/PUT` requests. But calling `.persist()` on the entity is not one of them. And that's rest-data-panache's fault for doing that blindly.

So we could do two things:

- Make sure rest-data-panache when it comes to being represented in OpenAPI, instructs it to not include the ID in `POST/PUT` for `PanacheEntity` entities. Perhaps @phillip-kruger has an idea of how we might do that?
- Make rest-data-panache ignore ID fields on `POST/PUT` by blanking them if the entities are instances of `PanacheEntity` (because if they have custom IDs set by extending `PanacheEntityBase` then the ID field has to be kept, if it is not auto-generated). But this would just silently ignore IDs passed by the users, which IMO is not the right thing to do.

Now that I think about it, the same problem would happen for users who extend `PanacheEntityBase` with their own auto-generated ID, but in this case the users would be able to add the `@Schema` annotation to make it read-only.

**Comment by nicolasduminil** [comment]
If by `rest-data-panache` you mean `quarkus-hibernate-orm-rest-data-panache`, then I'm not using it but `quarkus-hibernate-orm-panache`. And my entities are of the repository type, like:

    @ApplicationScoped
    public class CustomerRepository implements PanacheRepository<Customer> {}

I'm very surprised to face such an issue with such a basic thing that works since decades with Wildfly, JBoss, Payara, Glassfish, WebLogic and even Spring. And there doesn't seem to be any work-around. :-(

**Comment by nicolasduminil** [comment]
A reproducer is provided [here ](https://github.com/nicolasduminil/quarkus-swagger-issue)if it may help. The README.md file contains instructions of how to reproduce the issue.

**Comment by nicolasduminil** [comment]
Have added a separate branch, named `rest-data`, to the reproducer, which uses `quarkus-hibernate-orm-rest-data-panache` instead of `quarkus-hibernate-orm-panache`. The issue is present as well. No DTOs in this case.

**Comment by FroMage** [comment]
If you have your own DTO type, then it's up to you to generate or annotate it in a way that you don't accept `id` on POST/PUT if you don't support it. I don't see how it's a problem we can solve in this case. You could just do `customer.id = null` before you persist it if you can't have a DTO for POST/PUT that represents what you really accept (you don't accept an ID there, not really).

This problem is only relevant for `rest-data-panache` because here, users can't work around it (well, short of sending the right JSON on PUT/POST).

**Comment by nicolasduminil** [comment]
@FroMage : Not sure to understand ... 
"You could just do `customer.id = null` before you persist ...". I can't do anything since its Swagger which does everything. Once again, the issue, if it wasn't clear enough, is that when tested with Swagger, the REST API raises exceptions, while it works as expected when tested in @QuarkusTest tests.
"The problem is relevant for `rest-data-panache` ...". The problem is relevant for  `quarkus-hibernate-orm-rest-data-panache` but also for `quarkus-hibernate-orm-panache` as the reproducer clearly proves it.
It's funny to see that you're systematically asking for reproducers but, when such a reproducer is provided, at the price of several working hours, it is ignored.

**Comment by FroMage** [comment]
I'm not ignoring your reproducer, thanks for adding it, I read it all before I commented, sorry if it wasn't clear.

If you add `custom.id = null` right after this line: https://github.com/nicolasduminil/quarkus-swagger-issue/blob/master/src/main/java/fr/simplex_software/quarkus_swagger_issue/service/CustomerService.java#L38 it will just work. I'm saying this is your DTO which is wrong in accepting an ID when your code does not accept an ID because the ID is auto-generated. You can either fix your DTO to refuse that ID, or fix your service code to clear it up before you persist the entity.

My point is that in this case, this is code you wrote and maintain and can fix. You could even fix your entity so that swagger stops advertising that the ID is accepted, with `@Schema`.

This is not the case for the `rest-data-panache` extensions, where we need a solution to make Swagger stop advertising that ID is accepted when it's not. This part we do need to find a solution for.

**Comment by nicolasduminil** [comment]
Yes, I confirm that this way the `master` branch using  `quarkus-hibernate-orm-panache` works, thank you. I still don't understand why it works now and didn't work previously but it works. That's the 1st time that I need to add code in order to get Swagger tests working. The Quarkus tests don't need this statement to work, why does Swagger then ?
Additionally, this statement doesn't do anything because, printing the customer ID before executing it, it's already null. So why do I need to set to null a property which is already null ? Mystery :-).
The issue still stays for the `quarkus-hibernate-orm-rest-data-panache` but this is another story.
Thank you very much again for helping me getting the test case working, even if I'm still more confused than before :-).

**Comment by FroMage** [comment]
> The Quarkus tests don't need this statement to work, why does Swagger then ?
Additionally, this statement doesn't do anything because, printing the customer ID before executing it, it's already null. So why do I need to set to null a property which is already null ? Mystery :-).

Swagger adds `id = 0` to the examples, so you will get a non-null ID if your DTO accepts IDs. I'm not sure why it doesn't show up in your printing, but I strongly suspect it was not in fact `null`.

Your hand-written tests most likely do not set the entity ID so they are not affected, because they are correct, contrary to the Swagger-generated examples.

**Comment by nicolasduminil** [comment]
> ... I strongly suspect it was not null ...
Yeah, you're right, after careful verification, it appears that, as a matter of fact, it was 0, not null.
